### PR TITLE
Support RocksDB auto-tune rate limiter for the background IO

### LINKED ISSUE #1417: Add support of RocksDB auto-tune rate limiter for the background IO
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

https://rocksdb.org/blog/2017/12/18/17-auto-tuned-rate-limiter.html

### Solution
https://github.com/apache/incubator-kvrocks/blob/30e9fd788a4acb585566a008f4d5fff65ee06ce3/src/storage/storage.cc#L163

change to this
```c++
rocksdb::NewGenericRateLimiter(
        rate_limiter_bandwidth(),
        rate_limiter_refill_period_us(),
        rate_limiter_fairness(),
        rocksdb::RateLimiter::Mode::kWritesOnly,
        rate_limiter_auto_tuned()
      );
```

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by wanghenshui**
fix https://github.com/apache/incubator-kvrocks/issues/1417

### CODE REVIEW

File: src/storage/storage.cc [hunk 1]
```diff
@@ -159,8 +159,19 @@ rocksdb::Options Storage::InitRocksDBOptions() {
...
+            max_io_mb * static_cast<int64_t>(MiB),
+            100 * 1000, /* default */
+            10, /* default */
+            rocksdb::RateLimiter::Mode::kWritesOnly,
+            true
+          ));
+  } else {
+    rate_limiter_ =
       std::shared_ptr<rocksdb::RateLimiter>(rocksdb::NewGenericRateLimiter(max_io_mb * static_cast<int64_t>(MiB)));
+  }
```
Comment (PragmaTwice) [inline src/storage/storage.cc]:
```suggestion
  rate_limiter_ =
    std::shared_ptr<rocksdb::RateLimiter>(rocksdb::NewGenericRateLimiter(
          max_io_mb * static_cast<int64_t>(MiB),
          100 * 1000, /* default */
          10, /* default */
          rocksdb::RateLimiter::Mode::kWritesOnly,
          config_->rocks_db.rate_limiter_auto_tuned
        ));
```

**Comment by PragmaTwice** [comment]
To make the CI pass, you can run `./x.py format` in your local before pushing your code.

File: src/storage/storage.cc [hunk 2]
```diff
@@ -159,8 +159,19 @@ rocksdb::Options Storage::InitRocksDBOptions() {
   options.sst_file_manager = sst_file_manager_;
   int64_t max_io_mb = kIORateLimitMaxMb;
   if (config_->max_io_mb > 0) max_io_mb = config_->max_io_mb;
-  rate_limiter_ =
```
Comment (mapleFU) [inline src/storage/storage.cc:162]: What about:

```c++
    rate_limiter_ =
      std::shared_ptr<rocksdb::RateLimiter>(rocksdb::NewGenericRateLimiter(
            max_io_mb * static_cast<int64_t>(MiB),
            100 * 1000, /* default */
            10, /* default */
            rocksdb::RateLimiter::Mode::kWritesOnly,
            config_->rocks_db.rate_limiter_auto_tuned
          ));
```

Comment (wanghenshui) [reply src/storage/storage.cc:162]: sure

**Comment by wanghenshui** [comment]
@xiaobiaozhao  please benchmark this in `compaction with heavy write` scenario

**Comment by mapleFU** [comment]
This LGTM. Let wait some benchmarking for it

**Comment by xiaobiaozhao** [comment]
```
# The maximum allowed aggregated write rate of flush and compaction (in MB/s).
# If the rate exceeds max-io-mb, io will slow down.
# 0 is no limit
# Default: 500
max-io-mb 500
```
I think you can reuse this configuration 'max-io-mb'. If it is 0 use auto-tune.

**Comment by wanghenshui** [comment]
> ```
> # The maximum allowed aggregated write rate of flush and compaction (in MB/s).
> # If the rate exceeds max-io-mb, io will slow down.
> # 0 is no limit
> # Default: 500
> max-io-mb 500
> ```
> 
> I think you can reuse this configuration 'max-io-mb'. If it is 0 use auto-tune.

`auto tune` donesn't means auto detect write bandwidth, write bandwidth limit is always useful

File: tests/cppunit/config_test.cc [hunk 3]
```diff
@@ -79,6 +79,7 @@ TEST(Config, GetAndSet) {
       {"rocksdb.max_bytes_for_level_multiplier", "10"},
       {"rocksdb.level_compaction_dynamic_level_bytes", "yes"},
       {"rocksdb.max_background_jobs", "4"},
+
```
Comment (PragmaTwice) [inline tests/cppunit/config_test.cc]:
```suggestion
```
We do not expect newline without context : )

**Comment by git-hulk** [comment]
@xiaobiaozhao Did you test this PR?

**Comment by xiaobiaozhao** [comment]
> 



> @xiaobiaozhao Did you test this PR?
I will test it this weekend.

**Comment by xiaobiaozhao** [comment]
Press data into kvrocks script

> docker run -d --rm redislabs/memtier_benchmark:latest -t 4 -c 20 -s 172.27.255.244 -p 6666 --distinct-client-seed  --key-minimum=1000000000 --key-maximum=10000000000 --random-data  --key-prefix="" --data-size=50 -n 100000000 --pipeline=100 --command "mset  __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__ __key__ __data__"

write script
> docker run --rm redislabs/memtier_benchmark:latest -t 4 -c 20 -s 172.27.255.244 -p 6666 --distinct-client-seed  --key-minimum=1000000000 --key-maximum=10000000000 --key-prefix="" -n 1000 --command "set  __key__ __data__"

**Comment by mapleFU** [comment]
Thanks! Do you have idea that why the performance decline in normal p100(And does all performance optimization and decline comes from unstable test result)?

**Comment by xiaobiaozhao** [comment]
> Thanks! Do you have idea that why the performance decline in normal p100(And does all performance optimization and decline comes from unstable test result)?

I think this is bias of a test, p99 p9999 should be more convincing

**Comment by git-hulk** [comment]
This option may bring better performance benefits when the background compaction is heavy, not sure if the benchmark data was under this scenario. But anyway, it's good to merge since it also brings slight performance improvement.

**Review by xiaobiaozhao — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Comment by git-hulk** [comment]
Thanks all, merging...

**Pull request merged**